### PR TITLE
Add new stripe={...} prop

### LINKED ISSUE #8: Add support for a redux-stored `stripe` instance.
Perhaps something like:

```jsx
<Elements stripe={this.props.stripeInstance}>
   <MyCheckoutForm />
</Elements>
```

### LINKED ISSUE #24: Library does do not allow Stripe SDK to be loaded async
The provider expects `window.Stripe` to be set in its constructor. This forces the user to load the script synchronously (which is bad practice) _before_ they load their app JS (which is even worse).

Are there plans to rearchitect this library to support this?

I have a server-side rendered app, which does not include the Stripe SDK (since it's not needed on every page) -- on pages that require the SDK, we inject it & assume (perhaps slightly unsafely) that it has loaded by the time the user has entered their payment details.

Even ignoring the safety aspects, this approach is not possible with this library as it will throw an exception if the Stripe SDK is not present at first instantiation.

### LINKED ISSUE #64: SSR doesnt work
Server side rendering throw error

### LINKED ISSUE #120: Server side rendering (SSR) support
### Summary

Trying to use this library with SSR using [Next.js](https://github.com/zeit/next.js) and getting errors. What would it take to support SSR if it's not a fault of my own?

### Other information

I'm fairly new to React but I'm using Next.js v4 with React 16 to get SSR. Wrapping anything with `<StripeProvider>` results in an error: `ReferenceError: window is not defined`.

This appears to come from [Provider.js#L30](https://github.com/stripe/react-stripe-elements/blob/master/src/components/Provider.js#L30). Moving the bit of code concerned by `window` into `componentDidMount` didn't help (Elements.js is unhappy because it can't get the context in time from what I gather). Any chance to defer that once on the client?

I've put together a quick test case (stock Next.js install with a page that has `StripeProvider > Elements > CardElement`: https://github.com/MathieuLoutre/nextjs-stripe-elements

If you're not interested in SSR support, I'm also investigating triggering a re-render once on the client that includes the `StripeProvider` (but in that case it would be great if it could be scoped just to a part of the form and not the whole form!). I'll keep experimenting but I thought I'd flag this up as it's a great way to get a Stripe form up and running quickly and still be robust.

### PR DESCRIPTION
**Pull request by jez**
### Motivation

- "I don’t want to load Stripe.js until my customer needs it."
  - #24
- "I want to do the first render on the server for better performance."
  - #64, #120
- "I want to store my `stripe` instance in redux."
  - #8, #86 

### Summary

This adds a new `stripe={...}` prop to `<StripeProvider>`.
The new `stripe` prop is allowed to be null. The implication is that this
library is now designed to work in server-side and asynchronous environments.

See the README for the functionality enabled by this PR.

### Testing & documentation

The existing unit tests are brittle and so they fail.

The old demo works (<http://localhost:8080>)
There's a new async demo (<http://localhost:8080/async/>)

Any brittle tests of old functionality have been updated.
There are new tests for sync behavior, using `stripe` instead of `apiKey`.
There are new tests for async behavior.
There are new tests for SSR behavior.

### Commit summary

There are a lot of changes in this PR. To make it easier to understand,
consider reviewing each commit in sequence:

<details>
<summary>Click to expand</summary>

- **Allow either apiKey OR stripe param in Provider** (57b725f)

  Where to start? Look at:

  - `type State`: what's the state's shape look like each case?

  - the four if/then/else cases in the constructor: the "if both" and
  "if
   neither" cases arise when trying to use a product of values
   (`{apiKey, stripe}`) like a sum of values (`apiKey | stripe`).
   Half this code would be unnessary if JavaScript was typed statically.

  - `getChildContext`: notice how we either call the callback right now
   or stash it away for later.

- **Use new ProviderContext in Element/Elements** (ed5d43b)

  *All* the individual Element components will now act like they're
  **asynchronous**. `this.context.elements` is now *completely* replaced 
  by `this.context.addElementsListener`.

  This is possible for Elements because we never exposed the synchronous
  `this.props.elements` to the user.

  Meanwhile, this is *not* possible for injectStripe (see next commit) 
  because our API *did* expose the implicit synchronicity.

- **Use new Sync/Async ProviderContext in inject** (93577ef)

  The changes in this file actually gave me the most grief. The
  relatively few lines of code that needed to be changed in the end make
  me confident that we've arrived at an abstraction that elegantly
  solves the problem.

  Of note:

  - `type InjectedProps`: the injected `stripe` prop is now nullable,
   even if the programmer can *prove* that `stripe` will never be null.

  - We use `setState` within our callback to `addStripeLoadListener` to
   *force* a re-render of the wrapped component when the injected prop
   becomes available.

- **Use new changes in demo** (e5a46dc)

  In a world with no types, there are **no integration** changes needed.

  In a hypothetical world with types, this diff corresponds to the
  changes our users would have to deal with. Note how they have to check
  `if (this.props.stripe)` even though we can statically guarantee it
  will always be defined.

- **prettier** (fb4ab86)

  I un-prettier'd some previoius commits to make the diffs nicer.

- **Refactor some things** (28404db)

  - share more contextTypes
  - cut/paste requireElement after findElement (latter user the former)

- **Update brittle tests** (d876db9)

  These tests were testing implementation-specific details. Since some
  parts of the implementation changed, so have the tests.

  New functionality is NOT being tested by this commit.

- **Add tests for new surface area** (b8c071c)

  Provider: tests being given all combinations of 'apiKey' and 'stripe'

  inject: if Provider is async, ensure that we first give the wrapped 
  component `{stripe: null}` and later `{stripe: {...}}`.

  Element: Updating props runs error-free even when we haven't been
  given the Element (this._element) yet.

- **Whitespace only change** (a120286)

  We omitted indentation changes from the last commit to improve the
  diff.

- **Change webpack in anticipation of another demo** (3c94e83)

  Because the next commit files will be renamed, in this commit we can
  see what *actually* changed, rather than just a huge "file deleted"
  and another "file added".

- **Move main demo page to subfolder** (3e2b589)

- **Add new async demo and tests** (f3192cf)

  The async demo is copy/paste from the main demo. The only differences
  are *deletions*. This means no new code to review.

  It works by crafting and injecting a `<script>` tag into the page.

  It's served from http://localhost:8080/async/. In addition, it's
  "server-side" rendered in our tests (smoke test).

- **Cosmetic README source changes** (80986cd)

  These changes make the README play more nicely with my Markdown syntax 
  file.

- **Add documentation to README** (abc4f7e)

</details>

### Closes
  
Closes #8.
Closes #24.
Closes #64.
Closes #86.
Closes #120.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,10 @@ goes into more detail on the various customization options for Elements (e.g. st
   - [Setting up your payment form (`injectStripe`)](#setting-up-your-payment-form-injectstripe)
   - [Using individual `*Element` components](#using-individual-element-components)
   - [Using the `PaymentRequestButtonElement`](#using-the-paymentrequestbuttonelement)
+- [Alternative loading strategies](#alternative-loading-strategies)
```
Comment (michelle) [inline README.md]: I think this reference is out of date.

**Comment by michelle** [comment]
lgtm!

**Comment by williamccrobinson** [comment]
great timing! I was just coming across this issue as well 👍

**Pull request merged**